Crash: `AttributeError: 'If' object has no attribute 'name'`

### Bug description

Pylint crashes while parsing a project file

[pylint-crash-2022-01-08-14.txt](https://github.com/PyCQA/pylint/files/7833321/pylint-crash-2022-01-08-14.txt)


### Configuration

[tox.ini.txt](https://github.com/PyCQA/pylint/files/7833323/tox.ini.txt)

### Command used

```shell
pylint --rcfile /home/jdanek/repos/qpid/qpid-dispatch/cmake-build-debug/tests/tox.ini /home/jdanek/repos/qpid/qpid-dispatch/python /home/jdanek/repos/qpid/qpid-dispatch/console /home/jdanek/repos/qpid/qpid-dispatch/docs /home/jdanek/repos/qpid/qpid-dispatch/tests /home/jdanek/repos/qpid/qpid-dispatch/tools /home/jdanek/repos/qpid/qpid-dispatch/tools/qdstat /home/jdanek/repos/qpid/qpid-dispatch/tools/qdmanage
```


### Pylint output

```shell
/home/jdanek/repos/qpid/qpid-dispatch/tests/system_test.py:806:16: W0715: Exception arguments suggest string formatting might be intended (raising-format-tuple)
Exception on node <FunctionDef.assertRegex l.844 at 0x7fe7845bfe80> in file '/home/jdanek/repos/qpid/qpid-dispatch/tests/system_test.py'
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 72, in walk
    callback(astroid)
  File "/usr/lib/python3.10/site-packages/pylint/checkers/classes.py", line 1126, in visit_functiondef
    self._check_signature(node, parent_function, "overridden", klass)
  File "/usr/lib/python3.10/site-packages/pylint/checkers/classes.py", line 2022, in _check_signature
    f"{method1.parent.name}.{method1.name}",
AttributeError: 'If' object has no attribute 'name'
/home/jdanek/repos/qpid/qpid-dispatch/tests/system_test.py:1:0: F0001: Fatal error while checking '/home/jdanek/repos/qpid/qpid-dispatch/tests/system_test.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in '/home/jdanek/.cache/pylint/pylint-crash-2022-01-08-14.txt'. (fatal)
```


### Expected behavior

Pylint should not crash while parsing any file.

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.1
Python 3.10.1 (main, Dec 14 2021, 09:14:19) [GCC 11.2.1 20211209 releases/gcc-11.2.0-565-g99854374f9]
```


### OS / Environment

Fedora 35

### Additional dependencies

aiohttp==3.8.1
aiohttp-cors==0.7.0
aiosignal==1.2.0
alabaster==0.7.12
ansible==4.8.0
ansible-core==2.12.0
appdirs==1.4.4
arrow==1.2.1
asgiref==3.4.1
asn1crypto==1.4.0
astroid==2.9.1
astunparse==1.6.3
asv==0.4.2
async-timeout @ file:///builddir/build/BUILD/async-timeout-4.0.2/dist/async_timeout-4.0.2-py3-none-any.whl
atomicwrites==1.4.0
attrs @ file:///builddir/build/BUILD/attrs-21.4.0/dist/attrs-21.4.0-py2.py3-none-any.whl
awscli @ file:///builddir/build/BUILD/awscli-1.22.27/dist/awscli-1.22.27-py3-none-any.whl
Babel==2.9.1
backcall==0.2.0
backports.ssl_match_hostname==3.7.0.1
bcc==0.20.0
bcrypt==3.2.0
black @ file:///builddir/build/BUILD/black-21.12b0/dist/black-21.12b0-py3-none-any.whl
bleach==4.1.0
blinker==1.4
blockdiag==3.0.0
bokeh @ file:///builddir/build/BUILD/bokeh-2.4.2/dist/bokeh-2.4.2-py3-none-any.whl
boto==2.49.0
boto3==1.20.27
botocore==1.23.27
brotlipy==0.7.0
build==0.7.0
cached-property==1.5.2
cachetools @ file:///builddir/build/BUILD/cachetools-5.0.0/dist/cachetools-5.0.0-py3-none-any.whl
certifi==2021.10.8
cffi==1.15.0
chardet==4.0.0
charset-normalizer==2.0.9
check-manifest==0.47
chrome-gnome-shell==0.0.0
click==8.0.3
cloudpickle==2.0.0
colorama==0.4.4
commonmark==0.9.1
ConfigArgParse==1.5.3
coverage==6.2
createrepo-c==0.17.7
cryptography==35.0.0
cupshelpers==1.0
cxxfilt==0.3.0
cycler==0.11.0
Cython==0.29.26
darkflow==1.0.0
dask==2021.12.0
dataclasses==0.8
deap==1.3.0
decorator==5.1.0
defusedxml==0.7.1
distlib==0.3.4
distro==1.6.0
Django @ file:///builddir/build/BUILD/Django-4.0/dist/Django-4.0-py3-none-any.whl
docker==5.0.3
docker-pycreds==0.4.0
docopt==0.6.2
docutils==0.18.1
execnet==1.9.0
filelock @ file:///builddir/build/BUILD/filelock-3.4.2/dist/filelock-3.4.2-py3-none-any.whl
flake8==3.8.4
flake8-blind-except==0.2.0
flake8-class-newline==1.6.0
flake8-docstrings==1.6.0
Flask==2.0.2
frozenlist==1.2.0
fs==2.4.14
fsspec @ file:///builddir/build/BUILD/fsspec-2021.11.1/dist/fsspec-2021.11.1-py3-none-any.whl
funcparserlib @ file:///builddir/build/BUILD/funcparserlib-1.0.0a0/dist/funcparserlib-1.0.0a0-py2.py3-none-any.whl
funcsigs==1.0.2
future==0.18.2
gensim==4.1.2
Glances==3.2.3.1
google-auth==2.3.3
gpg===1.16.0-unknown
h5py==3.6.0
hacking==4.1.0
html5lib==1.1
icdiff==2.0.4
idna==3.3
imageio==2.13.5
imagesize==1.3.0
importlib-metadata @ file:///builddir/build/BUILD/importlib_metadata-4.10.0/dist/importlib_metadata-4.10.0-py3-none-any.whl
importlib-resources==5.4.0
iniconfig==1.1.1
intelhex==2.3.0
iotop==0.6
ipaddr==2.2.0
ipython==7.30.1
isc==2.0
isodate==0.6.1
isort==5.10.1
itsdangerous==2.0.1
jdcal==1.4.1
jedi==0.18.1
jeepney==0.7.1
Jinja2==3.0.3
jmespath==0.10.0
joblib==1.1.0
jsonpointer==2.2
jsonschema==3.2.0
kapidox==5.89.0
kcc==0.0.9
kerberos==1.3.1
keyring @ file:///builddir/build/BUILD/keyring-23.4.0/dist/keyring-23.4.0-py3-none-any.whl
kiwisolver==1.3.2
lazy-object-proxy==1.6.0
libcomps==0.1.18
libcst @ file:///builddir/build/BUILD/libcst-0.3.23/dist/libcst-0.3.23-py3-none-any.whl
libvirt-python==7.10.0
locket==0.2.1
lvm===2.02.186-2-.-2019-08-27-
lxml==4.7.1
MarkupSafe==2.0.1
matplotlib==3.4.3
matplotlib-inline==0.1.3
mccabe==0.6.1
meld==3.20.4
meson==0.60.3
mido==1.2.10
mock==4.0.3
MonkeyType==21.5.0
more-itertools @ file:///builddir/build/BUILD/more-itertools-8.12.0/dist/more_itertools-8.12.0-py3-none-any.whl
mpi4py==3.1.2
mpmath==1.2.1
multidict==5.2.0
mypy==0.930
mypy-extensions==0.4.3
netaddr==0.7.19
networkx==2.6.3
nftables==0.1
nose==1.3.7
nose-parameterized==0.6.0
ntlm-auth==1.5.0
numpy @ file:///builddir/build/BUILD/numpy-1.21.5/dist/numpy-1.21.5-cp310-cp310-linux_x86_64.whl
numpydoc==1.1.0
oauthlib==3.1.1
olefile==0.46
onnx==1.10.2
opcodes==0.3.14
ordereddict==1.1
packaging==21.3
pandas @ file:///builddir/build/BUILD/pandas-1.3.5/dist/pandas-1.3.5-cp310-cp310-linux_x86_64.whl
paramiko==2.4.2
parso==0.8.3
partd==1.2.0
pathlib2==2.3.6
pathspec==0.9.0
patsy==0.5.2
pbr==5.8.0
pep517==0.12.0
pexpect==4.8.0
pickleshare==0.7.5
Pillow==8.4.0
pkginfo==1.8.2
platformdirs @ file:///builddir/build/BUILD/platformdirs-2.4.1/dist/platformdirs-2.4.1-py3-none-any.whl
plotly @ file:///builddir/build/BUILD/plotly-5.5.0/dist/plotly-5.5.0-py2.py3-none-any.whl
pluggy==1.0.0
ply==3.11
posix-ipc==1.0.5
prettytable==2.5.0
prompt-toolkit==3.0.24
protobuf==3.18.1
psutil==5.9.0
ptyprocess==0.7.0
pwquality==1.4.4
py==1.11.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycairo==1.20.1
pycodestyle==2.6.0
pycparser==2.21
pydocstyle==6.1.1
pydot==1.4.2
pydot-ng==2.0.0
pydotplus==2.0.2
pyelftools==0.27
pyflakes==2.2.0
Pygments==2.11.1
PyGObject==3.42.0
pylint==2.12.2
PyMySQL==0.9.3
PyNaCl==1.4.0
pyOpenSSL==21.0.0
pyparsing==3.0.6
PyQt5==5.14.1.dev1912311030
Pyro4==4.82
pyroma==3.2
pyroute2==0.6.5
pyroute2.core==0.6.5
pyroute2.ethtool==0.6.5
pyroute2.ipdb==0.6.5
pyroute2.ipset==0.6.5
pyroute2.ndb==0.6.5
pyroute2.nftables==0.6.5
pyroute2.nslink==0.6.5
pyrsistent==0.18.0
PySocks==1.7.0
pytest==6.2.5
pytest-cov==3.0.0
pytest-repeat==0.9.1
pytest-rerunfailures==10.2
pytest-tap==3.3
python-dateutil==2.8.2
python-distutils-extra==2.39
python-dotenv==0.19.2
python-hwinfo==0.1.7
python-Levenshtein==0.12.2
pytz==2021.3
PyWavelets==1.2.0
PyYAML==6.0
readme-renderer==32.0
recommonmark==0.7.1
regex==2021.11.10
requests==2.27.0
requests-file==1.5.1
requests-kerberos==0.12.0
requests-ntlm==1.1.0
requests-oauthlib==1.3.0
requests-toolbelt==0.9.1
resolvelib==0.8.1
retype==21.12.0
rfc3986==1.5.0
rfc3987==1.3.8
rsa @ file:///builddir/build/BUILD/rsa-4.8/dist/rsa-4.8-py3-none-any.whl
ruamel.yaml==0.17.20
ruamel.yaml.clib==0.2.6
s3transfer==0.5.0
scikit-image @ file:///builddir/build/BUILD/scikit-image-0.19.1/dist/scikit_image-0.19.1-cp310-cp310-linux_x86_64.whl
scikit-learn==1.0.2
scipy @ file:///builddir/build/BUILD/scipy-1.7.3/dist/scipy-1.7.3-cp310-cp310-linux_x86_64.whl
scp==0.14.2
seaborn==0.11.2
SecretStorage==3.3.1
seqdiag==3.0.0
serpent==1.40
setuptools-scm==6.3.2
simplejson==3.17.6
six==1.16.0
smart-open==5.2.1
smartcols==0.3.0
snowballstemmer==2.2.0
sortedcontainers==2.4.0
Sphinx==4.3.2
sphinx-feature-classification==1.1.0
sphinx-rtd-theme==1.0.0
sphinxcontrib-apidoc==0.3.0
sphinxcontrib-applehelp==1.0.2
sphinxcontrib-blockdiag==3.0.0
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==2.0.0
sphinxcontrib-httpdomain==1.8.0
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-newsfeed==0.1.4
sphinxcontrib-programoutput==0.17
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-seqdiag==3.0.0
sphinxcontrib-serializinghtml==1.1.5
sphinxcontrib-websupport==1.2.4
sqlparse==0.4.2
SSSDConfig==2.6.0
statsmodels==0.13.0
sympy==1.8
tabulate==0.8.9
tap.py==3.1
templated-dictionary==1.1
tenacity==8.0.1
tflearn==0.5.0
Theano==1.0.5
threadpoolctl==3.0.0
tifffile==2021.11.2
toml==0.10.2
tomli @ file:///builddir/build/BUILD/tomli-2.0.0/dist/tomli-2.0.0-py3-none-any.whl
toolz==0.11.2
torch==1.9.0a0+gitunknown
tornado==6.1
tox @ file:///builddir/build/BUILD/tox-3.24.5/dist/tox-3.24.5-py2.py3-none-any.whl
tqdm==4.62.3
traitlets==5.1.1
twine @ file:///builddir/build/BUILD/twine-3.7.1/dist/twine-3.7.1-py3-none-any.whl
typed-ast==1.5.1
typing-inspect==0.7.1
typing_extensions @ file:///builddir/build/BUILD/typing_extensions-4.0.1/dist/typing_extensions-4.0.1-py3-none-any.whl
urllib3==1.26.7
virtualenv==20.13.0
voluptuous==0.12.2
watchdog==2.1.6
wcwidth==0.2.5
webcolors==1.11.1
webencodings==0.5.1
websocket-client==1.2.3
Werkzeug==2.0.2
Whoosh==2.7.4
wrapt==1.13.3
xmltodict==0.12.0
yarl==1.7.2
zeep==4.1.0
zipp @ file:///builddir/build/BUILD/zipp-3.7.0/dist/zipp-3.7.0-py3-none-any.whl
zope.event==4.5.0
zstandard==0.16.0
jdanek@clr-dev~/repos/qpid/qpid-dispatch/cmake-build-debug (originmaster) $ 2
pytz==2021.3
PyWavelets==1.2.0
PyYAML==6.0
readme-renderer==32.0
recommonmark==0.7.1
regex==2021.11.10
requests==2.27.0
requests-file==1.5.1
requests-kerberos==0.12.0
requests-ntlm==1.1.0
requests-oauthlib==1.3.0
requests-toolbelt==0.9.1
resolvelib==0.8.1
retype==21.12.0
rfc3986==1.5.0
rfc3987==1.3.8
rsa @ file:///builddir/build/BUILD/rsa-4.8/dist/rsa-4.8-py3-none-any.whl
ruamel.yaml==0.17.

--- a/ChangeLog
+++ b/ChangeLog
@@ -9,6 +9,12 @@ Release date: TBA
 ..
   Put new features here and also in 'doc/whatsnew/2.13.rst'
 
+
+* Fixed crash from ``arguments-differ`` and ``arguments-renamed`` when methods were
+  defined outside the top level of a class.
+
+  Closes #5648
+
 * Added several checkers to deal with unicode security issues
   (see `Trojan Sources <https://trojansource.codes/>`_ and
   `PEP 672 <https://www.python.org/dev/peps/pep-0672/>`_ for details) that also

--- a/doc/whatsnew/2.13.rst
+++ b/doc/whatsnew/2.13.rst
@@ -71,6 +71,11 @@ Extensions
 Other Changes
 =============
 
+* Fixed crash from ``arguments-differ`` and ``arguments-renamed`` when methods were
+  defined outside the top level of a class.
+
+  Closes #5648
+
 * Fixed false positive ``consider-using-dict-comprehension`` when creating a dict
   using a list of tuples where key AND value vary depending on the same condition.
 

--- a/pylint/checkers/classes/class_checker.py
+++ b/pylint/checkers/classes/class_checker.py
@@ -1325,7 +1325,7 @@ a metaclass class method.",
         ) and self._py38_plus:
             self.add_message(
                 "overridden-final-method",
-                args=(function_node.name, parent_function_node.parent.name),
+                args=(function_node.name, parent_function_node.parent.frame().name),
                 node=function_node,
             )
 
@@ -1997,24 +1997,24 @@ a metaclass class method.",
                     error_type = "arguments-differ"
                     msg_args = (
                         msg
-                        + f"was {total_args_refmethod} in '{refmethod.parent.name}.{refmethod.name}' and "
+                        + f"was {total_args_refmethod} in '{refmethod.parent.frame().name}.{refmethod.name}' and "
                         f"is now {total_args_method1} in",
                         class_type,
-                        f"{method1.parent.name}.{method1.name}",
+                        f"{method1.parent.frame().name}.{method1.name}",
                     )
                 elif "renamed" in msg:
                     error_type = "arguments-renamed"
                     msg_args = (
                         msg,
                         class_type,
-                        f"{method1.parent.name}.{method1.name}",
+                        f"{method1.parent.frame().name}.{method1.name}",
                     )
                 else:
                     error_type = "arguments-differ"
                     msg_args = (
                         msg,
                         class_type,
-                        f"{method1.parent.name}.{method1.name}",
+                        f"{method1.parent.frame().name}.{method1.name}",
                     )
                 self.add_message(error_type, args=msg_args, node=method1)
         elif (

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -1585,7 +1585,7 @@ accessed. Python regular expressions are accepted.",
             not isinstance(itemmethod, nodes.FunctionDef)
             or itemmethod.root().name != "builtins"
             or not itemmethod.parent
-            or itemmethod.parent.name not in SEQUENCE_TYPES
+            or itemmethod.parent.frame().name not in SEQUENCE_TYPES
         ):
             return None
 

--- a/tests/functional/a/arguments_renamed.py
+++ b/tests/functional/a/arguments_renamed.py
@@ -72,3 +72,29 @@ class ChildDefaults(ParentDefaults):
 
     def test3(self, dummy_param, arg2): # no error here
         print(f"arguments: {arg2}")
+
+# Check for crash on method definitions not at top level of class
+# https://github.com/PyCQA/pylint/issues/5648
+class FruitConditional:
+
+    define_eat = True
+
+    def brew(self, fruit_name: str):
+        print(f"Brewing a fruit named {fruit_name}")
+
+    if define_eat:
+        def eat_with_condiment(self, fruit_name:str, condiment: Condiment):
+            print(f"Eating a fruit named {fruit_name} with {condiment}")
+
+class FruitOverrideConditional(FruitConditional):
+
+    fruit = "orange"
+    override_condiment = True
+
+    if fruit == "orange":
+        def brew(self, orange_name: str): # [arguments-renamed]
+            print(f"Brewing an orange named {orange_name}")
+
+        if override_condiment:
+            def eat_with_condiment(self, fruit_name: str, condiment: Condiment, error: str): # [arguments-differ]
+                print(f"Eating a fruit named {fruit_name} with {condiment}")

--- a/tests/functional/a/arguments_renamed.txt
+++ b/tests/functional/a/arguments_renamed.txt
@@ -7,3 +7,5 @@ arguments-renamed:48:4:49:22:Child2.test:Parameter 'arg' has been renamed to 'va
 arguments-differ:51:4:52:58:Child2.kwargs_test:Number of parameters was 4 in 'Parent.kwargs_test' and is now 3 in overridden 'Child2.kwargs_test' method:UNDEFINED
 arguments-renamed:51:4:52:58:Child2.kwargs_test:Parameter 'var2' has been renamed to 'kw2' in overridden 'Child2.kwargs_test' method:UNDEFINED
 arguments-renamed:67:4:68:56:ChildDefaults.test1:Parameter 'barg' has been renamed to 'param2' in overridden 'ChildDefaults.test1' method:UNDEFINED
+arguments-renamed:95:8:96:59:FruitOverrideConditional.brew:Parameter 'fruit_name' has been renamed to 'orange_name' in overridden 'FruitOverrideConditional.brew' method:UNDEFINED
+arguments-differ:99:12:100:76:FruitOverrideConditional.eat_with_condiment:Number of parameters was 3 in 'FruitConditional.eat_with_condiment' and is now 4 in overridden 'FruitOverrideConditional.eat_with_condiment' method:UNDEFINED

--- a/tests/functional/o/overridden_final_method_py38.py
+++ b/tests/functional/o/overridden_final_method_py38.py
@@ -14,3 +14,18 @@ class Base:
 class Subclass(Base):
     def my_method(self): # [overridden-final-method]
         pass
+
+# Check for crash on method definitions not at top level of class
+# https://github.com/PyCQA/pylint/issues/5648
+class BaseConditional:
+
+    create_final_method = True
+    if create_final_method:
+        @final
+        def my_method(self):
+            pass
+
+class Subclass2(BaseConditional):
+
+    def my_method(self): # [overridden-final-method]
+        pass

--- a/tests/functional/o/overridden_final_method_py38.txt
+++ b/tests/functional/o/overridden_final_method_py38.txt
@@ -1,1 +1,2 @@
 overridden-final-method:15:4:16:12:Subclass.my_method:Method 'my_method' overrides a method decorated with typing.final which is defined in class 'Base':UNDEFINED
+overridden-final-method:30:4:31:12:Subclass2.my_method:Method 'my_method' overrides a method decorated with typing.final which is defined in class 'BaseConditional':UNDEFINED

